File: algorithm-kotlin/src/boj-2539.kt
import java.io.BufferedReader
import java.io.InputStreamReader

/**
 * 아니 이거 맞는데 코틀린 코드만 틀림.. 입력값이 뭔가 문제가 있는듯
 */
fun main() = with(BufferedReader(InputStreamReader(System.`in`))){
    val (R, C) = readLine().split(' ').map { it.toInt() }
    val PC = readLine().toInt()
    val WC = readLine().toInt()
    val colList = mutableListOf<Int>()
    var maxRow = 0
    repeat(WC) {
        val (r, c) = readLine().trim().split(' ').map { it.toInt() }
        maxRow = maxOf(maxRow, r)
        colList += c
    }
    colList.sort()
    fun decision(k : Int) : Boolean{
        var st = -1
        var cnt = 0
        colList.forEach {
            if(st == -1) {
                st = it
                cnt++
            }else if(it - st >= k) {
                st = it
                cnt++
            }
        }
        return cnt <= PC
    }
    fun optimize() : Int {
        var lo = minOf(maxRow, C) - 1
        var hi = maxOf(R, C) + 1
        while(lo + 1 < hi) {
            val mid = (lo + hi) / 2
            if(decision(mid))
                hi = mid
            else
                lo = mid
        }
        return hi
    }
    println(optimize())
}

File: etc/boj-2539.cc
#include <bits/stdc++.h>

using namespace std;

int main()
{
    int r, c;
    cin >> r >> c;
    int pc; cin >> pc;
    int wc; cin >> wc;
    vector<int> colList;
    int maxRow=0;
    for(int i=0;i<wc;++i){
        int a, b;cin >> a >> b;
        maxRow = max(maxRow, a);
        colList.push_back(b);
    }
    sort(colList.begin(), colList.end());
    auto decision = [&](int k) {
        int st = -1;
        int cnt = 0;
        for(auto& it : colList) {
            if(st == -1){
                st = it;
                cnt++;
            }else if(it - st >= k) {
                st = it;
                cnt++;
            }
        }
        return cnt <= pc;
    };
    int lo = min(maxRow,c)-1, hi = min(r, c);
    while(lo + 1 < hi) {
        int mid = (lo + hi) / 2;
        if(decision(mid))
            hi = mid;
        else    
            lo = mid;
    }
    cout << hi << endl;
    return 0;
}